Reformat code with `black`

sjdilkes opened the pull request:
Make sure code styling is consistent

daniel-mills-cqc commented:
Interesting, seems black is not rewriting too much with reassures me :P . A few comments:
- Will you reformat the tests?
- The main reason I moved to flake8 was that I found it easier to add automatic code formatting checks to the CI. Are there such checks in the case of black? This is the main issue for me as I find it a worse coding experience to enforce rules not enforced by the CI. At present we enforce PEP8 rules, and I'm not sure what more black insists upon?
- The notes on formatting in the README would also need updating if this is adopted repo wide.
- Did you mention that you are using a more up-to-date version of mypy locally? I see you made some changes to appease it. Could you update the version used by the CI?

As I think you'll have gathered I personally don't want to insist on using black. My own coding experience is made less pleasant by using a particular opinionated automatic code formatter than by checks of an unopinionated one, as we do now. I suppose this is a matter for democracy though so if there is a majority in favour of adopting the opinions of black then we can do that.

daniel-mills-cqc commented on qermit/__init__.py:
```diff
@@ -20,14 +20,9 @@
 from qermit.taskgraph.mitres import MitRes  # noqa:F401
 from qermit.taskgraph.mitex import MitEx  # noqa:F401
-from qermit.taskgraph.utils import SymbolsDict, MeasurementCircuit, ObservableTracker  # noqa:F401
+from qermit.taskgraph.utils import SymbolsDict, MeasurementCircuit, ObservableTracker # noqa:F401
```
Not meaning to get too into the weeds here, but this seems to contradict with PEP8. In particular "E261: at least two spaces before inline comment". I prefer to stick to PEP8 as I understand that it is the 'standard', although this could also be a version problem. Is there a way to make black be consistent here (see the imports above with two spaces for example)

sjdilkes replied:
Don't you worry, this was picked up by the CI & also was a user issue and not a black issue.

Black converted the import to 
```
from qermit.taskgraph.utils import (
 SymbolsDict,
 MeasurementCircuit```
 etc... which was completely at odds with PEP8 so I manually rejigged it.

daniel-mills-cqc commented on qermit/taskgraph/utils.py:
```diff
@@ -177,7 +177,7 @@ class MeasurementCircuit(object):
     """
 
-    def __init__(self, symbolic_circuit: Circuit, symbols: SymbolsDict = None):
+    def __init__(self, symbolic_circuit: Circuit, symbols: Optional[SymbolsDict] = None):
```
ah, is this what the newer mypy was picking up? Nice

sjdilkes replied:
A couple of cases of optionals + signal_filter not defining a method as abstract

daniel-mills-cqc reviewed (commented):
A couple of comment, mostly looking good. Good to see that there's not too much reformatting :P

sjdilkes commented:
> Interesting, seems black is not rewriting too much with reassures me :P . A few comments:
> 
> * Will you reformat the tests?
> * The main reason I moved to flake8 was that I found it easier to add automatic code formatting checks to the CI. Are there such checks in the case of black? This is the main issue for me as I find it a worse coding experience to enforce rules not enforced by the CI. At present we enforce PEP8 rules, and I'm not sure what more black insists upon?
> * The notes on formatting in the README would also need updating if this is adopted repo wide.
> * Did you mention that you are using a more up-to-date version of mypy locally? I see you made some changes to appease it. Could you update the version used by the CI?
> 
> As I think you'll have gathered I personally don't want to insist on using black. My own coding experience is made less pleasant by using a particular opinionated automatic code formatter than by checks of an unopinionated one, as we do now. I suppose this is a matter for democracy though so if there is a majority in favour of adopting the opinions of black then we can do that.

• Have just updated with reformatted tests
• `black --check .` will do it
• The code is still with in style guides enforced by README and CI so I don't think this needs updating - unless we formally switch to add `black` checks
• mypy version was hardcoded for some reason - I have removed the hardcoding

Consistent code style is good for readability amongst different developers. In this case consistent use of `black` makes our code consistent with the code style used over other Quantinuum repos such as `pytket`, `TKET` and the extension modules. We will have to disagree on coding experience as I'm fine with having my code auto-formatted into a consistent style.

I'm fairly easy on this - if you want to use `flake8` then that's cool with me. As I mentioned in another comment, `black` doesn't totally align with the current checking and I had to make some manual changes to make it align. I'll format any code I end up contributing to `qermit` using `black` and then fix CI issues to align with `flake8`. In general, as you said, the changes seem pretty minor/innocuous, mainly removing empty lines.

daniel-mills-cqc reviewed (approved):
Approved, but am I overlooking the change to the mypy requirements? Perhaps you could drop it in, else ignore me if I've overlooked it.

Looking good!

sjdilkes commented:
> Approved, but am I overlooking the change to the mypy requirements? Perhaps you could drop it in, else ignore me if I've overlooked it.
> 
> Looking good!

I _think_ the reason the `mypy` version wasn't updating is because requirements had the `mypy` version hardcoded to a specific version - a constraint I've now removed. Not sure why it was hardcoded as all the tests pass, I'm assuming a very outdated reason!
